sentinel/src/lib/change-retry-history.js

```diff
@@ -0,0 +1,36 @@
+const MAX_HISTORY = 1000;
+const MAX_RETRIES = 5;
+const historyKeys = [];
+const history = {};
+
+const getKey = (change) => change && `${change.id}${change.changes && change.changes[0] && change.changes[0].rev}`;
+
+const add = (change) => {
+  const key = getKey(change);
+  if (!key) {
+    return;
+  }
+
+  if (history[key]) {
+    history[key]++;
+  } else {
+    historyKeys.push(key);
+    history[key] = 1;
+  }
+
+  if (historyKeys.length > MAX_HISTORY) {
+    const deletedKey = historyKeys.shift();
+    delete history[deletedKey];
+  }
+};
+
+const shouldProcess = (change) => {
+  const key = getKey(change);
+
+  return !history[key] || history[key] <= MAX_RETRIES;
+};
+
+module.exports = {
+  add,
+  shouldProcess,
+};
```

Comment:
You know how much I love caches...

... but I think this is ok.

An alternative would be to only detect when the transition seq isn't advancing at all. That is to say if an entire batch is processed and 100% fails 5 times in a row then skip over them and move on. This way you're only caching 1 number instead of 1000 keys. Simpler and therefore less buggy?

One advantage of your approach is if you had 99 failing docs my approach would never get past them but yours would. 

Are there other advantages of your suggestion?

Reply:
I'll check to see how your suggestion looks like. I think it requires adding some conditionals in the `feed.js` code that would increase complexity.

Reply:
There's no guarantee that requesting changes from a certains seq will return the same results every time (because shards and because docs can get updated). 
So saying that a "batch" failed 5 times, it would still mean we need to check every doc in that batch and store which docs were in the failed batch. So we'd still be storing 100 changes in memory, along with the seq. 

Do you think this is worth pursuing and would be easier to maintain or understand than straight up caching doc tries without the seq?

Reply:
Yeah that's a good point. Go with it as is.